connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp

```diff
@@ -0,0 +1,1561 @@
+// Structure containing function pointers to the stack callbacks
+static radio_events_t *_radio_events;
+
+//SUBGHZ handle Structure definition
+SUBGHZ_HandleTypeDef hsubghz;
+ 
+// Data buffer used for both TX and RX
+// Size of this buffer is configurable via Mbed config system
+// Default is 255 bytes
+static uint8_t _data_buffer[MAX_DATA_BUFFER_SIZE_STM32WL];
+
+static uint8_t _operation_mode;
+static uint8_t _active_modem;
+    
+using namespace mbed;
+using namespace rtos;
+
+#ifdef MBED_CONF_RTOS_PRESENT
+/**
+ * Signals
+ */
+#define SIG_INTERRUPT        0x02
+#endif
+
+
+/**
+  * @brief voltage of vdd tcxo.
+  */
+#define TCXO_CTRL_VOLTAGE           TCXO_CTRL_1_7V
+
+/*!
+ * FSK bandwidth definition
+ */
+typedef struct {
+    uint32_t bandwidth;
+    uint8_t  register_value;
+} fsk_bw_t;
+
+static const fsk_bw_t fsk_bandwidths[] = {
+    { 4800, 0x1F },
+    { 5800, 0x17 },
+    { 7300, 0x0F },
+    { 9700, 0x1E },
+    { 11700, 0x16 },
+    { 14600, 0x0E },
+    { 19500, 0x1D },
+    { 23400, 0x15 },
+    { 29300, 0x0D },
+    { 39000, 0x1C },
+    { 46900, 0x14 },
+    { 58600, 0x0C },
+    { 78200, 0x1B },
+    { 93800, 0x13 },
+    { 117300, 0x0B },
+    { 156200, 0x1A },
+    { 187200, 0x12 },
+    { 234300, 0x0A },
+    { 312000, 0x19 },
+    { 373600, 0x11 },
+    { 467000, 0x09 },
+    { 500000, 0x00 }, // Invalid Bandwidth
+};
+
+const uint8_t sync_word[] = {0xC1, 0x94, 0xC1, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+// in ms                                 SF12    SF11    SF10    SF9    SF8    SF7
+const float lora_symbol_time[3][6] = {{ 32.768, 16.384, 8.192, 4.096, 2.048, 1.024 },  // 125 KHz
+    { 16.384, 8.192,  4.096, 2.048, 1.024, 0.512 },  // 250 KHz
+    { 8.192,  4.096,  2.048, 1.024, 0.512, 0.256 }
+}; // 500 KHz
+
+STM32WL_LoRaRadio::STM32WL_LoRaRadio(PinName rf_switch_ctrl1,
+									 PinName rf_switch_ctrl2,
+									 PinName rf_switch_ctrl3)
+    :
+        _rf_switch_ctrl1(rf_switch_ctrl1, PIN_OUTPUT, PullNone, 0),
+        _rf_switch_ctrl2(rf_switch_ctrl2, PIN_OUTPUT, PullNone, 0),
+        _rf_switch_ctrl3(rf_switch_ctrl3, PIN_OUTPUT, PullNone, 0)
+#ifdef MBED_CONF_RTOS_PRESENT
+    , irq_thread(osPriorityRealtime, 1024, NULL, "Thread_STM32WL")
+#endif
+		
+{
+    _radio_events = NULL;
+    _image_calibrated = false;
+    _force_image_calibration = false;
+    _active_modem = MODEM_LORA;
+#ifdef MBED_CONF_RTOS_PRESENT
+    irq_thread.start(callback(this, &STM32WL_LoRaRadio::rf_irq_task));
+#endif
+}
+
+STM32WL_LoRaRadio::~STM32WL_LoRaRadio()
+{
+
+}
+
+/**
+ * Acquire lock
+ */
+void STM32WL_LoRaRadio::lock(void)
+{
+    mutex.lock();
+}
+
+/**
+ * Release lock
+ */
+void STM32WL_LoRaRadio::unlock(void)
+{
+    mutex.unlock();
+}
+
+#ifdef MBED_CONF_RTOS_PRESENT
+/**
+ * Thread task handling IRQs
+ */
+void STM32WL_LoRaRadio::rf_irq_task(void)
+{
+    for (;;) {
+        uint32_t flags = ThisThread::flags_wait_any(0x7FFFFFFF);
+
+        lock();
+        if (flags & SIG_INTERRUPT) {
+            irq_process_isr();
```

Comment:
Not clear to me how this works ?
when flags has SIG_INTERRUPT you will call irq_process_isr that then set the flags_set(SIG_INTERRUPT);
Looks like a kind of deadlock ....
don't you rather want to call RadioIrqProcess here ?

Reply:
You are right, strange the deadlock didn't occur: corrected